hermes/probe/check_nil.go

```diff
@@ -0,0 +1,159 @@
+// TODO(evanSpendlove): Add exit status to return values.
+func CheckNilFile(ctx context.Context, target *Target, client *stiface.Client, logger *logger.Logger) (m.ExitStatus, error) {
+	bucket := target.Target.GetBucketName()
+
+	// TODO(evanSpendlove): Find the Nil file on the storage system, alert if not present.
+	query := &storage.Query{Prefix: "Hermes_NIL"}
+	start := time.Now()
+	objIter := (*client).Bucket(bucket).Objects(ctx, query) // Can add API call timeouts here.
+	end := time.Now()
+	var nilFilename string
+	for {
+		obj, err := objIter.Next()
+		if err != nil {
+			target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_LIST_FILES]][m.ExitStatuses[m.FILE_MISSING]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+			return fmt.Errorf("checkNilFile.list_files_failed: unable to find nil file: %w", err)
+		}
+		nilFilename = obj.Name
+	}
+	target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_LIST_FILES]][m.ExitStatuses[m.SUCCESS]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+	logger.Infof("checkNilFile: nil file located on target storage system.")
+
+	// TODO(evanSpendlove): Get the contents of the NIL file.
+	nilFile := (*client).Bucket(bucket).Object(nilFilename)
+
+	start = time.Now()
+	reader, err := nilFile.NewReader(ctx)
+	end = time.Now()
+	if err != nil {
+		target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_GET_FILE]][m.ExitStatuses[m.FILE_READ_FAILURE]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+		return fmt.Errorf("checkNilFile.read_file_failed: unable to read nil file: %w", err)
+	}
+	target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_GET_FILE]][m.ExitStatuses[m.SUCCESS]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+	logger.Infof("checkNilFile: nil file gotten from target storage system.")
+
+	// TODO(evanSpendlove): Amend metric recording here to use api_READ or probe op read and verify metrics.
+	start = time.Now()
+	nilContents, err := ioutil.ReadAll(reader)
+	if err != nil {
+		target.LatencyMetrics.ProbeOpLatency[m.ProbeOps[m.READ_FILE]][m.ExitStatuses[m.FILE_READ_FAILURE]].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
+		return fmt.Errorf("checkNilFile.read_file_failed: unable to read nil file: %w", err)
+	}
+
+	logger.Infof("checkNilFile: nil file read complete.")
+
+	// Verify checksum
+	// TODO(evanSpendlove): Re-evaluate how best to time verifying a file, possibly adding hashing time to other verification.
+	hashTime := time.Now()
+	hash := sha1.New()
+	if _, err := io.Copy(hash, reader); err != nil {
+		target.LatencyMetrics.ProbeOplatency[m.ProbeOps[m.VERIFY_FILE]][m.ExitStatuses[m.FILE_CORRUPTED]].Metric("latency").AddFloat64(time.Now().Sub(hashTime).Seconds())
+		return nil, fmt.Errorf("checkNilFile.compute_checksum_failed: unable to compute checksum of file contents: %w", err)
+	}
+	if fmt.Sprintf("%x", hash) != nilFilename[len("Hermes_NIL_"):] {
```

Comment:
This looks brittle. I would implement a `NilFile` type with a `CheckSum()` method that returns `[]byte` and would compare that with `hash`.

Reply:
I've refactored this to use a `nilFile` struct with a `validateChecksum()` method.

We still need to check the checksum against the checksum in the name, however, as this is part of validating that the file contents have not changed.